### PR TITLE
3180 deprecate rejectonnotfound

### LINKED ISSUE #3180: Deprecate `rejectOnNotFound` and  document new concept [ Prisma 4 ]
Part of the Prisma 4 effort: Docs should mention that `rejectOnNotFound` is a deprecated concept and point / document the new way of doing it.

Related issue: https://github.com/prisma/prisma/issues/12159

Andrew's notes:
SWE - Serhii

We'll need changes to:
* https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#rejectonnotfound - add deprecation notice on existing material, add new method

See this thread for discussion about the adopted solution: https://prisma-company.slack.com/archives/C02FZENHD4N/p1654883538432409

This feature currently applies to findUnique and findFirst
Will the new methods allow custom errors?
The new methods are called findUniqueOrThrow()  and findFirstOrThrow()
invlude - description, remarks (if appropriate), options, return type, reference, examples.

Remove rejectOnNotFound from fidUnique and findFirst - or rather, mark as deprecated, and link to new methods.

The solution will be 
* Chosen methods names are findUniqueOrThrow()  and findFirstOrThrow()
* ...instead of `findFirst(…, rejectOnNotFound: true)`, and so on

**Breaking change**
In the upgrade guide, we should document:
* Upgrade path for non-typescript users
* Upgrade path for typescript users

Done and merged

### PR DESCRIPTION
**Pull request by andrew-walford-prisma**
## Describe this PR

* Added deprecation notices for `rejectonnotfound`
* Added docs for `findFirstOrThrow` and `findUniqueOrThrow`

### CODE REVIEW

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 1]
```diff
@@ -374,6 +374,12 @@ const prisma = new PrismaClient({
 
 ### <inlinecode>rejectOnNotFound</inlinecode>
 
+<Admonition type="info">
+
+**Deprecated:** `rejectOnNotFound` is deprecated in v4.0.0. From v4.0.0, use the queries [`findUniqueOrThrow`](#inlinecodefinduniqueorthrowinlinecode) or [`findFirstOrThrow`](#inlinecodefindfirstorthrowinlinecode).
```
Comment (SevInf) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Maybe worth noting that it's also scheduled for removal in 5.0.0

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 2]
```diff
@@ -552,23 +559,42 @@ const result = await prisma.user.findUnique({
...
+We introduced `findUniqueOrThrow` in v4.0.0. It replaces the [`rejectOnNotFound`](#inlinecoderejectonnotfoundinlinecode) option. `rejectOnNotFound` is deprecated in v4.0.0.
+
+</Admonition>
+
+`findUniqueOrThrow` retrieves a single data record in the same way as [`findUnique`](#inlinecodefinduniqueinlinecode). However, if the query does not find a record, it returns `NotFoundError: No User found error`.
+
+`findUniqueOrThrow` differs from `findUnique` as follows:
+
+- Its return type is non-nullable. For example, `post.findUnique()` can return `post` or `null`, but `post.findUniqueOrThrow()` always returns `post`.
+- It is not compatible with the [`$transaction` API](/concepts/components/prisma-client/transactions#the-transaction-api).
```
Comment (SevInf) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: It is only incompatible with "array of promises" style API, interactive transactions work as expected. In fact, I'd suggest interactive transactions as a workaround. 

Maybe also worth mentioning in what way exactly it is incompatible: `NotFoundError`, thrown by these methods won't rollback the transaction. If query fails for any other reason on the database side, transaction will roll back normally.

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 1 above)
Comment (SevInf) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
**Deprecated:** `rejectOnNotFound` is deprecated in v4.0.0. From v4.0.0, use the methods [`findUniqueOrThrow`](#inlinecodefinduniqueorthrowinlinecode) or [`findFirstOrThrow`](#inlinecodefindfirstorthrowinlinecode).
```

Or do we normally call them "queries" throughout the docs?

**Review by SevInf — CHANGES_REQUESTED** [review]
Left a few comments (mostly on `findUnique` part, but same applies for `findFirst` as well), overall looks good.

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 1 above)
Comment (andrew-walford-prisma) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: We might not always be consistent, unfortunately, but we do call them "model queries" in this doc and "queries" elsewhere in the docs, so I'll stick with the terminology I used.

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 2 above)
Comment (andrew-walford-prisma) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Thanks - I'll add this detail.

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx (hunk 1 above)
Comment (andrew-walford-prisma) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: For the sake of docs maintainability, we'll omit this information (the version might change, and then we'll likely end up with incorrect information in the docs)

**Review by SevInf — APPROVED** [review]

**Review by dpetrick — APPROVED** [review]

**Pull request merged**